homeassistant/helpers/event.py

```diff
@@ -354,36 +356,315 @@ def remove_listener() -> None:
+    old_state
+        The old state of the entity that changed.
+    new_state
+        New state of the entity that changed.
+
+    Parameters
+    ----------
+    hass
+        Home assistant object.
+    template
+        The template to calculate.
+    action
+        Callable to call with results. See above for arguments.
+    variables
+        Variables to pass to the template.
+
+    Returns
+    -------
+    Callable to unregister the listener.
 
-    # Local variable to keep track of if the action has already been triggered
-    already_triggered = False
+    """
 
     @callback
-    def template_condition_listener(entity_id: str, from_s: State, to_s: State) -> None:
+    def state_changed_listener(
+        event: Event,
+        template: Template,
+        last_result: Optional[str],
+        result: Union[str, TemplateError],
+    ) -> None:
         """Check if condition is correct and run action."""
-        nonlocal already_triggered
-        template_result = condition.async_template(hass, template, variables)
-
-        # Check to see if template returns true
-        if template_result and not already_triggered:
-            already_triggered = True
-            hass.async_run_job(action, entity_id, from_s, to_s)
-        elif not template_result:
-            already_triggered = False
-
-    return async_track_state_change(
-        hass, template.extract_entities(variables), template_condition_listener
+        if isinstance(result, TemplateError):
+            _LOGGER.exception(result)
+            return
+
+        if result_as_boolean(last_result) or not result_as_boolean(result):
+            return
+
+        hass.async_run_job(
+            action,
+            event.data.get("entity_id"),
+            event.data.get("old_state"),
+            event.data.get("new_state"),
+        )
+
+    info = async_track_template_result(
+        hass, template, state_changed_listener, variables
     )
 
+    return info.async_remove
+
 
 track_template = threaded_listener_factory(async_track_template)
 
 
+_UNCHANGED = object()
+
+
+class TrackTemplateResultInfo:
+    """Handle removal / refresh of tracker."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        template: Template,
+        action: Callable,
+        variables: Optional[TemplateVarsType],
+    ):
+        """Handle removal / refresh of tracker init."""
+        self.hass = hass
+        self._template = template
+        self._action = action
+        self._variables = variables
+        self._last_result: Optional[str] = None
+        self._last_exception = False
+        self._all_listener: Optional[Callable] = None
+        self._domains_listener: Optional[Callable] = None
+        self._entities_listener: Optional[Callable] = None
+        self._info = template.async_render_to_info(variables)
+        if self._info.exception:
+            self._last_exception = True
+            _LOGGER.exception(self._info.exception)
```

Comment:
Normally we try to avoid side effects in init method and put those in a setup method instead.

Reply:
I can make  async_track_template_result do that.  We should probably make the class protected as well as we really don't want external usage of it directly.

Reply:
Updated here https://github.com/home-assistant/core/pull/38934